src/erlang_types/dnf_ty_function.erl

```diff
@@ -37,17 +37,54 @@ is_empty_coclause_corec(AllPos, Neg, T, M) ->
       BigSTuple = lists:foldl(fun(FunTy, Acc) ->
         ty_rec:union(Acc, domains_to_tuple(ty_function:domains(FunTy)))
                               end, domains_to_tuple(ty_function:domains(P)), []),
+      % TODO: Fix me
       is_empty_cont_corec(BigSTuple, AllPos, Neg, M);
-    {[P | Pos], Neg, _} ->
-      % TODO here do these simplifications
-      % A -> B && C -> B == A|C -> B
-      % A -> B && A -> C == A -> B&C
-      BigSTuple = lists:foldl(fun(FunTy, Acc) ->
-        ty_rec:union(Acc, domains_to_tuple(ty_function:domains(FunTy)))
-                              end, domains_to_tuple(ty_function:domains(P)), Pos),
+    {Pos, Neg, _} ->
+      % Transform list of function types to tuples
+      NewPos = [{domains_to_tuple(ty_function:domains(P)), ty_function:codomain(P)} || P <- Pos],
+
+      % As an optimization merge positive function types
+      [{Domains1, _} | Pos1] = merge_pos(NewPos),
+
+      BigSTuple = lists:foldl(fun({Domains, _}, Acc) -> ty_rec:union(Acc, Domains) end, Domains1, Pos1),
       is_empty_cont_corec(BigSTuple, AllPos, Neg, M)
   end.
 
+% Assume list contains tuples (domain_tuple, codomain)
+% Merges positive function types according to following simplification rules:
+% (A -> B) ∧ (A -> C) = (A -> B ∧ C)
+% (A -> C) ∧ (B -> C) = (A ∨ B -> C)
+merge_pos([]) -> [];
+merge_pos([P | Rest]) ->
+  NewPos = merge_pos_fun(P, Rest),
+  % TODO: Find a better way to check that a merge happened
+  MergedPos = case Rest == NewPos of
+    % Nothing changed, continue with the next function
+    true -> [P | merge_pos(Rest)];
+    % List changed, i.e. merge happened. Don't need P
+    _ ->
+      %io:format(user, "Functions merged!~n", []),
+      merge_pos(NewPos)
+  end,
+  MergedPos.
+
+% Merges a single function type with a list of function types
+merge_pos_fun(_, []) -> [];
+merge_pos_fun({Domains1, Codomain1}, Pos) ->
+  % For now just check if the list has the same length
+  lists:map(fun({Domains2, Codomain2}) ->
+    case {ty_rec:is_equivalent(Domains1, Domains2), ty_rec:is_equivalent(Codomain1, Codomain2)} of
```

Comment:
Why can we do semantic checks during function simplification here? Is it guaranteed that we don't loop because we're doing checks on a smaller type?